### PR TITLE
os/linux/elf: avoid using ldd for listing dynamic dependencies

### PR DESCRIPTION
**Pull request by alebcay**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Current approach is to query `ldd` and capture its output (which includes resolved paths to shared libraries), then filters based on the needed libraries determined through checking the dynamic section.

New approach is to search for each needed library listed in the dynamic section, trying to locate it via runpath/rpath and the system `ld.so.conf` paths.

With the new approach, we avoid passing (potentially untrusted) files to `ldd`. Instead, we analyze the ELF file statically (inspect the dynamic section) and only try to perform lookups for the shared libraries indicated in the dynamic section.

This change is intended to modify how the dynamic dependencies lookup happens, but should not alter any internal or public-facing API (only private methods/classes changed) - for instance, the output semantics, format, and type of `Pathname("/some/path/to/an/executable").dynamically_linked_libraries` should remain unchanged.

### CODE REVIEW

File: Library/Homebrew/os/linux/elf.rb [hunk 1]
```diff
@@ -157,6 +145,49 @@ def needed_libraries_using_patchelf_rb(path)
       patcher = path.patchelf_patcher
       [patcher.soname, patcher.needed]
     end
+
+    def ld_so_conf(conf_path = "/etc/ld.so.conf")
```
Comment (Bo98) [inline Library/Homebrew/os/linux/elf.rb:149]: You will probably need to consider `HOMEBREW_PREFIX/etc/ld.so.conf` if Homebrew glibc is used

Comment (alebcay) [reply Library/Homebrew/os/linux/elf.rb:149]: > You will probably need to consider `HOMEBREW_PREFIX/etc/ld.so.conf` if Homebrew glibc is used

In the case that it is, should `/etc/ld.so.conf` still be considered (after considering `HOMEBREW_PREFIX/etc/ld.so.conf`) or should only `HOMEBREW_PREFIX/etc/ld.so.conf` be considered?

Comment (Bo98) [reply Library/Homebrew/os/linux/elf.rb:149]: It replaces.

`HOMEBREW_PREFIX/lib/ld.so` should point to the correct dynamic linker being used. You can run `HOMEBREW_PREFIX/lib/ld.so --list-diagnostics` and extract `path.sysconfdir` to get the `etc` directory.

Also relevant is `path.system_dirs`, which are used if nothing in the config files were found, _except_ if the executable sets `DF_1_NODEFLIB` in `DT_FLAGS_1` in which case not only should this step be skipped but anything in the config that have those directories as a prefix should also be skipped.

Comment (alebcay) [reply Library/Homebrew/os/linux/elf.rb]: > `HOMEBREW_PREFIX/lib/ld.so` should symlink to the correct dynamic linker being used whether that be system or Homebrew glibc. You can run `HOMEBREW_PREFIX/lib/ld.so --list-diagnostics` and extract `path.sysconfdir` to get the `etc` directory.

We now execute `HOMEBREW_PREFIX/lib/ld.so --list-diagnostics` to get `path.sysconfdir`. If `HOMEBREW_PREFIX/lib/ld.so` does not exist (for instance, on a fresh `brew` install before any other install command is executed - e.g. the initial state of the `brew` container), the fallback config directory `/etc` is used.

> Also relevant is path.system_dirs, which are used if nothing in the config files were found, except if the executable sets DF_1_NODEFLIB in DT_FLAGS_1 in which case not only should this step be skipped but anything in the config that have those directories as a prefix should also be skipped.

I _think_ I've designed this as intended, hopefully I understood the last part correctly ("anything in the config that have those directories as a prefix should also be skipped" >> "ignore any paths specified in `ld.so.conf.d` files that are (direct or indirect) children of any path listed in `path.system_dirs`".

File: Library/Homebrew/os/linux/elf.rb [hunk 2]
```diff
@@ -157,6 +147,63 @@ def needed_libraries_using_patchelf_rb(path)
...
+    def dt_flags_1(path)
+      readelf = DevelopmentTools.locate "readelf"
+      readelf_output = Utils.popen_read(readelf, "--dynamic", path.expand_path.to_s)
+      return [] unless $CHILD_STATUS.success?
+
+      match = readelf_output.match(/^\s*0x0*6ffffffb\s+\(FLAGS_1\)\s+Flags:\s*(.*)$/)
+      return [] unless match
+
+      match.captures.compact.first.split(/\s+/)
+    end
```
[the remark this answers is not in the document]
Comment (alebcay) [reply Library/Homebrew/os/linux/elf.rb]: Nice, didn't know that `patchelf.rb` had `segment_by_type` and `tag_by_type` (and also didn't know about `ELFTools`, that's very helpful).

File: Library/Homebrew/os/linux/ld.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+      sig { returns(T::Array[String]) }
+      def system_dirs
+        system_dirs = []
+
+        brewed_ld_so = HOMEBREW_PREFIX/"lib/ld.so"
+        return system_dirs unless brewed_ld_so.exist?
+
+        ld_so_output = Utils.popen_read(brewed_ld_so, "--list-diagnostics").split("\n")
+        return system_dirs unless $CHILD_STATUS.success?
```
Comment (Bo98) [inline Library/Homebrew/os/linux/ld.rb:36]: This is duplicated between the two - maybe worth extracting into another function.

**Review by Bo98 — APPROVED** [review]
Looks good!

Just as a couple questions:
* Have you tested this on a few formulae?
* What's the performance difference? Faster/slower and by how much?

**Comment by alebcay** [comment]
So far I've spot-tested a few binaries/libraries, like `zsh` binary from the `zsh` formula and `libleptonica.so.6` from `leptonica`, to confirm that the output of `Pathname(<path to ELF file>).dynamically_linked_libraries` is the same. Open to any ideas on how to test further, especially for covering the case where `DF_1_NODEFLIB` is set.

As far as performance, the performance seems to be about the same. Testing with the following:

```ruby
def snippet()= 100.times { Pathname("/home/linuxbrew/.linuxbrew/Cellar/leptonica/1.84.1/lib/libleptonica.so.6").dynamically_linked_libraries }
Benchmark.measure { snippet }
```

Before:
```
=> #<Benchmark::Tms:0x00000040084902a8
 @cstime=1.1855609999999928,
 @cutime=7.002927,
 @label="",
 @real=13.041561631999684,
 @stime=0.21336699999999809,
 @total=11.345018999999986,
 @utime=2.943163999999996>
 ```
 
 After:
 ```
 => #<Benchmark::Tms:0x0000004008601da8
 @cstime=1.219255,
 @cutime=7.094854,
 @label="",
 @real=13.061064489999808,
 @stime=0.18731199999999998,
 @total=11.323314,
 @utime=2.821893>
```

The above "after" result is with no caching mechanism implemented.

As mentioned earlier, there is possibly room for improvements here:
- Cache the output of `ld.so --list-diagnostics`
  - Per `brew` invocation, i.e. memoize (don't persist to disk)
  - Persistent cache - need to figure out what is the right interval for invalidation (this feels a bit overkill)
- Read/parse/use `ld.so.cache` before trying to check library paths (and memoize/cache this)

File: Library/Homebrew/extend/pathname.rb [hunk 4]
```diff
@@ -490,6 +490,12 @@ def zipinfo
                  .encode(Encoding::UTF_8, invalid: :replace)
                  .split("\n")
   end
+
+  sig { params(parent: Pathname).returns(T::Boolean) }
+  def child_of?(parent)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/pathname.rb:495]: Can we avoid adding more monkey patches to `Pathname` in favour of e.g. a new `Utils` method? CC @dduugg for thoughts.

File: Library/Homebrew/os/linux/ld.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+module OS
+  module Linux
+    # Helper functions for querying `ld` information.
+    #
+    # @api private
+    module Ld
+      module_function
+
+      sig { returns(String) }
+      def brewed_ld_so_diagnostics
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux/ld.rb:13]: Would be nice to have some unit tests for this module maybe?

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work so far, a few suggested tweaks, feel free to push back on any.

File: Library/Homebrew/os/linux/ld.rb (hunk 5 above)
Comment (alebcay) [reply Library/Homebrew/os/linux/ld.rb:13]: Added a unit test particularly for `library_paths` which is parsing the `ld.so.conf` files (IMO, the least trivial part of the module).

It made more sense to provide one big input (containing multiple corner cases) for one test case and then confirm that all the expected corner cases are reflected in the output - let me know if you have a better/more idiomatic approach in mind.

File: Library/Homebrew/os/linux/ld.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# typed: true
```
Comment (dduugg) [inline Library/Homebrew/os/linux/ld.rb:1]: Let's add a `sig` for `library_paths` and make this `typed: strict`

File: Library/Homebrew/os/linux/ld.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+# frozen_string_literal: true
+
+module OS
+  module Linux
+    # Helper functions for querying `ld` information.
+    #
+    # @api private
+    module Ld
+      module_function
```
Comment (dduugg) [inline Library/Homebrew/os/linux/ld.rb:10]: I recommend removing this, making the methods class defs (`def self.`), and removing the `.rbi` file.

File: Library/Homebrew/utils/path.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# typed: true
+# frozen_string_literal: true
+
+module Utils
+  module Path
+    module_function
```
Comment (dduugg) [inline Library/Homebrew/utils/path.rb:6]: Similar to above: I recommend removing this, making the methods class defs (`def self.`), removing the `.rbi` file, and making this `typed: strict`. (I'd like to have aspects of this enforced by rubocop, but there are a number of existing violations to clean up.)

**Review by dduugg — CHANGES_REQUESTED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks again @alebcay!
